x/muxer/main.go

```diff
@@ -0,0 +1,167 @@
+//
+// Most rpc methods are resolved via a schema lookup based precomputed mapping created
+// when the Muxer is initialized, with earlier servers getting priority.
+//
+// For example:
+//
+//	Given server s1 serving resources r1, r2 and server s2 serving r1, r3, the muxer
+//	m1 := Mux(host, s1, s2) will dispatch r1 and r2 to s1. m1 will dispatch only r3 to
+//	s2.  If we swap the order of of creation: m2 := Mux(host, s2, s1) we see different
+//	prioritization. m2 will serve r1 and r3 to s2, only serving r2 to s1.
+//
+// Most methods are fully dispatch based:
+//
+//   - Create, Read, Update, Delete, Check, Diff: The type is extracted from the URN
+//     associated with the request. The server who's schema provided the resource is routed
+//     the whole request.
+//
+//   - Construct: The type token is passed directly. The server who's schema provided the
+//     resource is routed the whole request.
+//
+//   - Invoke, StreamInvoke, Call: The type token is passed directly. The server who's
+//     schema provided the function is routed the whole request.
+//
+// Each provider specifies in it's schema what options it accepts as configuration. Config
+// based endpoints filter the schema so each provider is only shown keys that it expects
+// to see. It is possible for multiple subsidiary providers to accept the same key.
+//
+//   - CheckConfig: Broadcast to each server. Any diffs between returned results errors.
+//
+//   - DiffConfig: Broadcast to each server. Results are then merged with the most drastic
+//     action dominating.
+//
+//   - Configure: Broadcast to each server for individual configuration. When computing
+//     the returned set of capabilities, each option is set to the AND of the subsidiary
+//     servers. This means that the Muxed server is only as capable as the least capable
+//     of its subsidiaries.
+//
+// A dispatch strategy doesn't make sense for methods related to the provider as a
+// whole. The following methods are broadcast to all providers:
+//
+//   - Cancel: Each server receives a cancel request.
+//
+// The remaining methods are treated specially by the Muxed server:
+//
+//   - GetSchema: When Mux is called, GetSchema is called once on each server. The muxer
+//     merges each schema with earlier servers overriding later servers. The origin of each
+//     resource and function in the presented schema is remembered and used to route later
+//     resource and function requests.
+//
+//   - Attach: `Attach` is never called on Muxed providers. Instead the host passed into
+//     `Mux` is replaced. If subsidiary servers where constructed with the same `host` as
+//     passed to `Mux`, then they will observe the new `host` spurred by `Attach`.
+//
+//   - GetMapping: TODO. GetMapping returns an opaque blob. It's not clear how to merge
+//     blobs. We might require that only one provider responds with a mapping, and just
+//     forward the first provider that responds.
+type Main struct {
```

Comment:
This is not a great name for the API. Since we're in experimental its' ok we'll fix it up later.